Mathlib/LinearAlgebra/RootSystem/OfBilinear.lean

```diff
@@ -0,0 +1,186 @@
+/-
+Copyright (c) 2024 Scott Carnahan. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Scott Carnahan
+-/
+import Mathlib.LinearAlgebra.RootSystem.Defs
+
+/-!
+# Root pairings made from bilinear forms
+A common construction of root systems is given by taking the set of all vectors in an integral
+lattice for which reflection yields an automorphism of the lattice.  In this file, we generalize
+this construction, replacing the ring of integers with an arbitrary commutative ring and the
+integral lattice with an arbitrary reflexive module equipped with a bilinear form.
+
+## Main definitions:
+ * `IsReflective`: Length is a regular value of `R`, and reflection is definable.
+ * `coroot_of_reflective`: The coroot corresponding to a reflective vector.
+ * `of_Bilinear`: The root pairing whose roots are reflective vectors.
+
+## TODO
+ * properties
+-/
+
+open Set Function Module
+
+noncomputable section
+
+variable {ι R M N : Type*} [CommRing R] [AddCommGroup M] [Module R M] [AddCommGroup N] [Module R N]
+
+namespace RootPairing
+
+section Construction
+
+/-- A vector `x` is reflective with respect to a bilinear form if multiplication by its norm is
+injective, and for any other vector `y`, there is a scalar that takes the norm of `x` to twice the
+inner product of `x` and `y`. These conditions are what we need when describing reflection as a map
+taking `y` to `y - 2 • (B x y) / (B x x) • x`. -/
+def IsReflective (B : M →ₗ[R] M →ₗ[R] R) (x : M) : Prop :=
+    IsRegular (B x x) ∧ ∀y : M, ∃r : R, B x x * r = 2 * B x y
```

Comment:
How about:
```suggestion
structure IsReflective (B : M →ₗ[R] M →ₗ[R] R) (x : M) : Prop :=
  regular : IsRegular (B x x)
  dvd_two_mul : ∀ y, B x x ∣ 2 * B x y
```

Reply:
I also think we should use `variable` to save a few characters
```lean
variable (B : M →ₗ[R] M →ₗ[R] R) (x : M)
```
and maybe also add
```lean
lemma isReflective_of_dvd_two [IsDomain R] [NeZero (2 : R)] (x : M) (hx : B x x ∣ 2) :
    IsReflective B x where
  regular := isRegular_of_ne_zero <| fun contra ↦ by simp [contra, two_ne_zero (α := R)] at hx
  dvd_two_mul y := hx.mul_right (B x y)
```